test/storage/local_file_source.test.cpp

```diff
@@ -0,0 +1,101 @@
+#include <mbgl/storage/local_file_source.hpp>
+#include <mbgl/platform/platform.hpp>
+#include <mbgl/util/run_loop.hpp>
+
+#include <unistd.h>
+#include <limits.h>
+#include <gtest/gtest.h>
+
+namespace {
+
+std::string toAbsoluteUrl(const std::string& fileName) {
+    char buff[PATH_MAX + 1];
+    char* cwd = getcwd( buff, PATH_MAX + 1 );
+    return "file:://" + std::string(cwd) + "/test/fixtures/storage/assets/" + fileName;
+}
+
+} // namespace
+
+using namespace mbgl;
+
+TEST(LocalFileSource, EmptyFile) {
+    util::RunLoop loop;
+
+    LocalFileSource fs;
+
+    std::unique_ptr<AsyncRequest> req = fs.request({ Resource::Unknown, toAbsoluteUrl("empty") }, [&](Response res) {
+        req.reset();
+        EXPECT_EQ(nullptr, res.error);
+        ASSERT_TRUE(res.data.get());
+        EXPECT_EQ("", *res.data);
+        loop.stop();
+    });
+
+    loop.run();
+}
+
+TEST(LocalFileSource, NonEmptyFile) {
+    util::RunLoop loop;
+
+    LocalFileSource fs;
+
+    std::unique_ptr<AsyncRequest> req = fs.request({ Resource::Unknown, toAbsoluteUrl("nonempty") }, [&](Response res) {
+        req.reset();
+        EXPECT_EQ(nullptr, res.error);
+        ASSERT_TRUE(res.data.get());
+        EXPECT_EQ("content is here\n", *res.data);
+        loop.stop();
+    });
+
+    loop.run();
+}
+
+TEST(LocalFileSource, NonExistentFile) {
+    util::RunLoop loop;
+
+    LocalFileSource fs;
+
+    std::unique_ptr<AsyncRequest> req = fs.request({ Resource::Unknown, toAbsoluteUrl("does_not_exist") }, [&](Response res) {
+        req.reset();
+        ASSERT_NE(nullptr, res.error);
+        EXPECT_EQ(Response::Error::Reason::NotFound, res.error->reason);
+        ASSERT_FALSE(res.data.get());
+        // Do not assert on platform-specific error message.
+        loop.stop();
+    });
+
+    loop.run();
+}
+
+TEST(LocalFileSource, ReadDirectory) {
+    util::RunLoop loop;
+
+    LocalFileSource fs;
+
+    std::unique_ptr<AsyncRequest> req = fs.request({ Resource::Unknown, toAbsoluteUrl("directory") }, [&](Response res) {
+        req.reset();
+        ASSERT_NE(nullptr, res.error);
+        EXPECT_EQ(Response::Error::Reason::NotFound, res.error->reason);
+        ASSERT_FALSE(res.data.get());
+        // Do not assert on platform-specific error message.
+        loop.stop();
+    });
+
+    loop.run();
+}
+
+TEST(LocalFileSource, URLEncoding) {
```

Comment:
For the sake we are no missing anything, can you make a test with a extremely large file name, like `memset(filename, 'x', PATH_MAX);`

Wondering what would happen if you try to read something you don't have permission to (like `/root/foo.txt`).